EKF/ekf.cpp

```diff
@@ -160,42 +160,63 @@ bool Ekf::initialiseFilter()
 	// Sum the magnetometer measurements
 	if (_mag_buffer.pop_first_older_than(_imu_sample_delayed.time_us, &_mag_sample_delayed)) {
 		if (_mag_sample_delayed.time_us != 0) {
-			_mag_counter ++;
-
-			// wait for all bad initial data to be flushed
-			if (_mag_counter <= uint8_t(_obs_buffer_length + 1)) {
+			if (_mag_counter == 0) {
 				_mag_lpf.reset(_mag_sample_delayed.mag);
 
 			} else {
 				_mag_lpf.update(_mag_sample_delayed.mag);
 			}
+
+			_mag_counter++;
 		}
 	}
 
 	// accumulate enough height measurements to be confident in the quality of the data
 	if (_baro_buffer.pop_first_older_than(_imu_sample_delayed.time_us, &_baro_sample_delayed)) {
 		if (_baro_sample_delayed.time_us != 0) {
-			_baro_counter ++;
-
-			// wait for all bad initial data to be flushed
-			if (_baro_counter <= uint8_t(_obs_buffer_length + 1)) {
-				_baro_hgt_offset = _baro_sample_delayed.hgt;
+			if (_baro_counter == 0) {
+				_baro_hgt_offset  = _baro_sample_delayed.hgt;
 
-			} else if (_baro_counter > (uint8_t)(_obs_buffer_length + 1)) {
+			} else {
 				_baro_hgt_offset = 0.9f * _baro_hgt_offset + 0.1f * _baro_sample_delayed.hgt;
 			}
+
+			_baro_counter++;
 		}
 	}
 
-	const bool not_enough_baro_samples_accumulated = _baro_counter <= 2u * _obs_buffer_length;
-	const bool not_enough_mag_samples_accumulated = _mag_counter <= 2u * _obs_buffer_length;
-
-	if (not_enough_baro_samples_accumulated || not_enough_mag_samples_accumulated) {
-		return false;
+	if (_params.mag_fusion_type <= MAG_FUSE_TYPE_3D) {
+		if (_mag_counter < _obs_buffer_length) {
```

Comment:
@dagar can you confirm that the derivers for these sensors will not publish until valid data is available. In the past we have had the situation where some sensors could publish zero before the sensor had properly started. From memory the 2xbuffer was a hack to work around this.

Reply:
That's the goal, and if there are still any offenders I'd like to find and fix them.

We're now very careful to not just fully reset a sensor, but also to give them their full reset time as per the datasheet. Then on transfer wherever possible sanity check the returned data before publishing (status bits, data ready, overflow, etc).

 - https://github.com/PX4/PX4-Autopilot/blob/149ac16bb442fc64eef320794a306b755758b5d4/src/drivers/magnetometer/isentek/ist8308/IST8308.cpp#L111
 - https://github.com/PX4/PX4-Autopilot/blob/149ac16bb442fc64eef320794a306b755758b5d4/src/drivers/magnetometer/isentek/ist8308/IST8308.cpp#L176